### PR TITLE
Add code generation for interface and union enums

### PR DESCRIPTION
**Pull request by designatednerd**
NOTE: Draft until the JSON here gets finalized.

In Swift Codegen, we're going to be using enums to represent the possible types of unions and interfaces. This PR takes information returned from the server about what interface and union types are used and uses them to generate those enums. 

Also threw in an additional test on regular enums since we didn't have one covering modifiers.

### CODE REVIEW

File: Sources/ApolloCodegenLib/ASTOutput.swift [hunk 1]
```diff
@@ -8,6 +8,12 @@ struct ASTOutput: Codable, Equatable {
   /// An array of all fragments to generate code for.
   let fragments: [ASTFragment]
   
-  /// An array of "all" types used <-- TODO: Figure out why some are not in here
```
Comment (amadeu01) [inline Sources/ApolloCodegenLib/ASTOutput.swift:11]: Sorry the ignorance, but have you figured it out why some of the types weren't there before?

Comment (designatednerd) [reply Sources/ApolloCodegenLib/ASTOutput.swift:11]: Yes, excellent question! 

The answer is that `typesUsed` was designed to be only for _Input_ types, though it was not really documented as such. Unions and interfaces were not included, and I couldn't figure out why when I initially started working through the JSON, which is why I left myself this `TODO`. 

In an upcoming version of code generation, union and interface types will be included - this PR works off some temporary JSON to support generating code for that.

**Comment by designatednerd** [comment]
Just a quick note that things are still being finalized with this, I'll be picking back up with the codegen once this is merged.

**Comment by designatednerd** [comment]
OK! We're *finally* in business on this. Merging away!

**Pull request merged**